src/components/browser-viewport/browser-viewport.vue

```diff
@@ -1,15 +1,63 @@
 <template>
-  <iframe :src="url" class="browser-viewport">
-    <p>Your browser does not support iframes.</p>
-  </iframe>
+    <iframe v-if="isFrameable" :src="url" class="browser-viewport"></iframe>
+    <img v-else-if="imageUrl" :alt="`Screenshot of ${url}`" :src="imageUrl" class="browser-viewport">
+    <div v-else class="browser-viewport"><p>Site doesn't allow embedding. Fetching as image...</p></div>
 </template>
 
 <script>
+  import fetchHeaders from '../../lib/fetch-headers'
+  import fetchScreenshot from '../../lib/fetch-screenshot'
   import VueTypes from 'vue-types'
 
   export default {
     props: {
       url: VueTypes.string.isRequired
+    },
+    data () {
+      return {
+        imageUrl: undefined,
+        isFrameable: true
+      }
+    },
+    mounted: function () {
+      this.checkFrameOptions(this.url)
+    },
+    methods: {
+      checkFrameOptions: function (url) {
+        this.isFrameable = true
+        fetchHeaders(url)
+          .then(headers => this.setFrameable(url, !headers['x-frame-options']))
+          .catch(this.showError)
+      },
+      getDimensions: function () {
+        const dimensions = this.$el.getBoundingClientRect()
+        const dpr = window.devicePixelRatio
+        const width = Math.ceil(dimensions.width * dpr)
+        const height = Math.ceil(dimensions.height * dpr)
+        return { width, height }
+      },
+      fetchScreenshot: function (url) {
+        const { width, height } = this.getDimensions()
+        this.imageUrl = undefined
+        fetchScreenshot({ url, width, height })
```

Comment:
fetchScreenshot is the same name as this.fetchScreenshot, but it's a different function. I found that a bit confusing.